services/QuillLMS/spec/services/classlink_integration/enrollment_persister_spec.rb

```diff
@@ -0,0 +1,156 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ClasslinkIntegration::EnrollmentPersister do
+  let(:incoming_student_enrollment) do
+    {
+      role: User::STUDENT,
+      user_sourced_id: 'student-1',
+      classroom_sourced_id: 'classroom-1',
+      status: 'active'
+    }
+  end
+
+  let(:incoming_teacher_enrollment) do
+    {
+      role: User::TEACHER,
+      user_sourced_id: 'teacher-1',
+      classroom_sourced_id: 'classroom-1',
+      status: 'active'
+    }
+  end
+
+  let(:classlink_instance) { create(:classlink_instance) }
+
+  describe '#run' do
+    context 'no user or classroom found' do
+      it 'should do nothing and return early' do
+        expect do
+          described_class.new(
+            incoming_enrollment: incoming_teacher_enrollment,
+            existing_provider_users: [],
+            existing_provider_classrooms: []
+          ).run
+        end
+          .to not_change(ClassroomsTeacher, :count)
+          .and not_change(StudentsClassrooms, :count)
+      end
+    end
+
+    context 'association already exists' do
+      let(:classroom1) { create(:classroom) }
+      let!(:classlink_classroom1) do
+        create(
+          :classlink_classroom,
+          classroom: classroom1,
+          provider_instance: classlink_instance,
+          external_id: incoming_student_enrollment[:classroom_sourced_id]
+        )
+      end
+      let(:classroom2) { create(:classroom) }
+      let!(:classlink_classroom2) do
+        create(
+          :classlink_classroom,
+          classroom: classroom2,
+          provider_instance: classlink_instance
+        )
+      end
+
+      context 'role: teacher' do
```

Comment:
This context block should be outside of `#run` since it's not calling `run` but `create_classroom_teacher`